hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java

```diff
@@ -50,25 +50,97 @@ public static RecordInfo buildRecordInfo(Record record) {
+                    //exceptional case: the cachedValue is not locked, it just contains an instance of Thread.
+                    //this can happen when user put an instance of Thread into a map
+                    //(=it should never happen, but never say never...)
+                    return cachedValue;
+                }
+                //it looks like some other thread actually locked the cachedValue
+                Thread.yield();
+            } else if (cachedValue == null) {
+                Object fromCache = tryToLoadIntoCache(record, serializationService);
+                if (fromCache != null) {
+                    return fromCache;
+                }
+            } else {
+                //it's not the 'in-progress' marker && it's not a null -> it has to be actual cachedValue
+                return cachedValue;
             }
+            cachedValue = record.getCachedValue();
         }
-        return value;
     }
 
-    private static boolean shouldCache(Record record, Object value) {
-        boolean isCachableRecordType = record instanceof CachedDataRecordWithStats || record instanceof CachedDataRecord;
-        if (!isCachableRecordType) {
-            return false;
+    private static Object tryToLoadIntoCache(Record record, SerializationService serializationService) {
+        Object valueBeforeCas = record.getValue();
+        if (!shouldCache(valueBeforeCas)) {
+            return valueBeforeCas;
         }
+
+        Thread currentThread = Thread.currentThread();
+        if (record.casCachedValue(null, currentThread)) {
+            //we managed to lock the record for ourselves
+            Object valueAfterCas = record.getValue();
+            if (valueAfterCas == valueBeforeCas) {
+                //this check is needed to make sure a partition thread had not changed the value
+                //before we won the CAS
+                Object object = serializationService.toObject(valueAfterCas);
+                object = ThreadWrapper.wrapIfNeeded(object);
+                if (record.casCachedValue(currentThread, object)) {
+                    return object;
+                }
+            } else {
+                //the value has changed -> let's remove the lock.
+                //we have to CAS the lock out as it could had been already removed by the partition thread
+                record.casCachedValue(currentThread, null);
+            }
+        }
+        return null;
+    }
+
+    static boolean shouldCache(Object value) {
         return value instanceof Data && !((Data) value).isPortable();
     }
 
+
+    /**
+     * currentThread inside cachedValue acts as "deserialization in-progress" marker
+     * if the actual deserialized value is instance of Thread then we need to wrap it
+     * otherwise it might be mistaken for the "deserialization in-progress" marker.
+     *
+     */
+    private static final class ThreadWrapper extends Thread {
+        private final Thread delegate;
```

Comment:
changed to `storedThread`. It's still not a great name, but it has very small scope.